src/client/Client.cc

```diff
@@ -8086,8 +8086,12 @@ int Client::_do_setattr(Inode *in, struct ceph_statx *stx, int mask,
     std::string path;
     if (make_absolute_path_string(in, path)) {
       ldout(cct, 20) << " absolute path: " << path << dendl;
-      if (path.length())
-        path = path.substr(1);    // drop leading /
+
+      // drop any leading / (it might contain leadin '//')
+      while (path.length() && path[0] == '/') {
+        path = path.substr(1);
```

Comment:
The func that can take some modification can be in `client/Dentry.h`:
https://github.com/ceph/ceph/blob/3956c4278abb8f3f97e1c48924bf741e64a68d82/src/client/Dentry.h#L80-L94

Watch out for `s += "/";`, it unconditionally adds a `/` as prefix; this is the func being called by `Client::make_absolute_path_string` and other callers within MDS code. Fixing this would make sure we never end up at this bug again ever. What do you guys think?

Reply:
> This should be part of `make_absolute_path_string`, no? All of the callers are modifying the `path` afterwards to remove leading `/`. Please refactor it and make it consistent.

The 'make_absolute_path_string'  is currently returning a path with  leading '//' instead of '/'. Fixing it to return withe leading singel '/' is one way to fix. Even then, all the callers of `make_absolute_path_string` need to remove the remaining '/' which they do currently.

But why should this be a caller's responsibility ? Shall we transfer the responsibility to `bool MDSCapMatch::match_path(string_view target_path)` ? where the actual mds auth caps path comparison happens.

Background:
1. The constructor of `MDSCapMatch` always removes leading '//' of the mds auth caps' path that is set as part of `fs authorize`. This is the path against which the `target_path` is compared with. The normalize_path does it.

```
  MDSCapMatch(const std::string& fsname_, const std::string& path_,                                                            
              bool root_squash_, int64_t uid_=MDS_AUTH_UID_ANY,                                                                
              const std::vector<gid_t>& gids_={}) {                                                                            
    fs_name = fsname_;                                                                                                         
    path = path_;                                                                                                              
    root_squash = root_squash_;                                                                                                
    uid = (uid_ == 0) ? -1 : uid_;                                                                                             
    gids = gids_;                                                                                                              
                                                                                                                               
    normalize_path();                                                                                                          
  } 
```
2. So now, it is the responsibilty of the caller of the `MDSCapMatch::match_path` to remove the leading '//' of the target_path. There are two callers.

     A). server/mds: The `Session:check_access` constructs the path from server inode using `in->make_path_string` which  I think gives a path with single '/' prefix. That is removed later in `Session::check_access` and sent down to call `MDSCapMatch::match_path`

    ```
    int Session::check_access(CInode *in, unsigned mask,
                              int caller_uid, int caller_gid,
                              const vector<uint64_t> *caller_gid_list,
                              int new_uid, int new_gid)
      {
      string path;
      CInode *diri = NULL;
      if (!in->is_base())
       diri = in->get_projected_parent_dn()->get_dir()->get_inode();
      if (diri && diri->is_stray()){
        path = in->get_projected_inode()->stray_prior_path;
        dout(20) << __func__ << " stray_prior_path " << path << dendl;
      } else {
        in->make_path_string(path, true);
        dout(20) << __func__ << " path " << path << dendl;
      }
      if (path.length())
        path = path.substr(1);    // drop leading /
      ....
      ....
    ```

    B) client: The `open` and `setattr` constructs the path using the client inode using `in->make_absolute_path_string` which returns the path with '//' and then removes the single '/'  before calling 'MDSCapMatch::match_path' function. This is the bug.
    
 Possible Fixes:
1. Make the `make_absolute_path_string` to return with single '/' and let callers of `MDSCapMatch::match_path' takes care of removing the remaining '/'. The code for which is already in place.
2. Remove the leading '//'  of target_path in 'MDSCapMatch::match_path' itself. Clients need not worry about removing.
3. Remove the leading '//' in `Client::mds_check_access` which calls the 'MDSCapMatch::match_path'. This is cleaner since all the callers of `mds_check_access` need not take care of removing leading '/'. This can be done at one place.

If option 2 doesn't have any issues, I think it's cleaner.

@vshankar @batrick Let me now your thoughts.

Reply:
> The func that can take some modification can be in `client/Dentry.h`:
> 
> https://github.com/ceph/ceph/blob/3956c4278abb8f3f97e1c48924bf741e64a68d82/src/client/Dentry.h#L80-L94
> 
> Watch out for `s += "/";`, it unconditionally adds a `/` as prefix; this is the func being called by `Client::make_absolute_path_string` and other callers within MDS code. Fixing this would make sure we never end up at this bug again ever. What do you guys think?

@dparmar18 I think it's a side effect of recursive nature of the function.

Reply:
I have fixed with option 1 and option 3 both. Let me know your thoughts.

Reply:
Yes, that's right!